### PR TITLE
KEP-8522: Add time-based same-priority preemption proposal

### PR DESCRIPTION
**Pull request by sohankunkerkar**
#### What type of PR is this?
/kind documentation
/kind feature

#### Which issue(s) this PR fixes:

Related to #8522 

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

**Comment by sohankunkerkar** [comment]
cc @Ladicle @dddwsd @mimowo

**Comment by Ladicle** [comment]
Thanks for the PR! I have a few questions regarding the behavior when Hierarchical Cohorts (HC) are enabled.

I'm curious if the following scenarios are covered in the current scope:
- Cross-CQ Reclaim: Can a workload in CQ-A preempt a workload of the same priority in CQ-B, provided the target has exceeded its minAdmitDuration?
- Within-CQ Protection: To avoid unnecessary churn within the same team, can we protect a new workload that does not preempt an existing one within the same CQ (e.g., CQ-B), even if minAdmitDuration has passed?

**Comment by sohankunkerkar** [comment]
>Cross-CQ Reclaim: Can a workload in CQ-A preempt a workload of the same priority in CQ-B, provided the target has exceeded its minAdmitDuration?

No, this is explicitly out of scope for this KEP. The `minAdmitDuration` configuration only applies to `withinClusterQueue` preemption. Cross-CQ preemption (including reclaim scenarios in Hierarchical Cohorts) continues to be operated by existing Fair Sharing algorithms and the `reclaimWithinCohort`/`borrowWithinCohort` policies.

 If cross-CQ time-based protection is needed, the semantics would require careful design. A few open questions:                                                             
                                                                                                                                                                             
1. Whose `minAdmitDuration` applies? If CQ-A (minAdmitDuration: 2h) is reclaiming from CQ-B (minAdmitDuration: 4h), which threshold governs whether CQ-B's workload is protected?                                                                                                                                                                 
2. Should reclaim of own quota be subject to `minAdmitDuration`? If CQ-A lent resources to CQ-B, should CQ-A have to wait before reclaiming its own nominal quota?           
3. Where would this be configured? On the Cohort? On the borrowing CQ? On reclaimWithinCohort?                                                                             
                                                                                                                                                                             
If you have a concrete use case for cross-CQ time-based protection, I'd love to hear more details.

>Within-CQ Protection: To avoid unnecessary churn within the same team, can we protect a new workload that does not preempt an existing one within the same CQ (e.g., CQ-B), even if minAdmitDuration has passed?

 Yes, the new workload is protected by its own `minAdmitDuration`. Each workload's protection is based on its own quota reservation time, not on how other workloads in the CQ are doing. So the same-priority workload can be preempted if either:                                                                                                                                                       
  1. It was admitted after the preemptor was created (existing LowerOrNewerEqualPriority behavior)                                                                           
  2. It has exceeded `minAdmitDuration` (new behavior)

**Comment by Ladicle** [comment]
Thanks for the reply!

In our environment, each project has its own ClusterQueue (CQ), which shares resources via a Hierarchical Cohort. Here's our scenarios:
- Cross-CQ (Fairness): To ensure fairness, we want to prevent a single project from monopolizing shared resources for too long. Therefore, we would like to allow cross-CQ preemption for same-priority workloads once they have exceeded their `minAdmitDuration`.
- Within-CQ (Stability/FIFO): On the other hand, within a single project (the same CQ), we prefer FIFO-like behavior. We want to avoid preempting workloads within the same CQ to minimize unnecessary stops, even if they have already passed the `minAdmitDuration` threshold.

> 3. Where would this be configured? On the Cohort? On the borrowing CQ? On reclaimWithinCohort?

For our specific use case, sharing `minAdmitDuration` at the Cohort level would be enough.

File: keps/8522-time-based-same-priority-preemption/README.md [hunk 1]
```diff
@@ -0,0 +1,353 @@
...
+    WithinClusterQueue  PreemptionPolicy    `json:"withinClusterQueue,omitempty"`
+
+    // withinClusterQueueConfig provides additional configuration for the
+    // withinClusterQueue preemption policy. Only valid when withinClusterQueue
+    // is set to LowerOrNewerEqualPriority.
+    // +optional
+    WithinClusterQueueConfig *WithinClusterQueueConfig `json:"withinClusterQueueConfig,omitempty"`
+}
+
+type WithinClusterQueueConfig struct {
```
Comment (kannon92) [inline keps/8522-time-based-same-priority-preemption/README.md:206]: I don't think this needs to be solved for this KEP but I have another use case actually.

I was discussing with a customer that they would like a way to control preemption based on how long it has been admitted.

Say if a job has run for 24 hours it should no longer be considered with preemption after it hits 24 hours.

Do you think it could be possible to consider this in design or at least think through how these policies could work together?

**Review by kannon92 — COMMENTED** [review]
I had one minor item but I think this is a well thought through proposal and a good addition to the project.

**Review by mimowo — COMMENTED** [review]
It is an important feature, but also complex, and the timeline for the timeline for 0.17 is very close. I would prefer to move it 0.18. You are free to start prototyping to de-risk the 0.18 release.

File: keps/8522-time-based-same-priority-preemption/README.md [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+this can lead to:
+
+- One team's long-running workload blocking other teams indefinitely
+- Users submitting artificially low resource requests to get admitted first,
+  then consuming resources for extended periods
+- Reduced cluster utilization efficiency due to suboptimal scheduling
+
+### Goals
+
+- Extend the `LowerOrNewerEqualPriority` preemption policy with a
```
Comment (mukund-wayve) [inline keps/8522-time-based-same-priority-preemption/README.md:77]: Hi Sohan, this is really useful for us and an important feature we are after. Thanks for writing this KEP!

But we have a use case where it would help to have two separate durations instead of one, because there are two different situations:

`minAdmitDuration` — for workloads that were already in the queue BEFORE the pending workload arrived. They got in first, so they should get a longer guaranteed runtime (e.g., 4h).

`opportunisticMinAdmitDuration` — for workloads that arrived AFTER the pending workload but got admitted first because they were small enough to fit. Today these can be preempted right away; this field would let them run for a short time first (e.g., 20m) so they can finish or make progress.

With a single duration, we can't handle both cases well: if we set it high, small jobs that jumped the queue are protected for too long; if we set it low, jobs that were rightfully first don't get enough time.

Example:

```
withinClusterQueueConfig:
  minAdmitDuration: 2h
  opportunisticMinAdmitDuration: 20m
```
WorkloadA (3 CPU) is admitted at T=0. WorkloadB (3 CPU) arrives at T=1m but can't fit, so it waits. WorkloadC (1 CPU) arrives at T=2m and fits in the remaining space, so it gets admitted. With this config, C can be preempted after 20m (it jumped the queue) and A after 2h (it was there first).

---

An user story to capture this could be:

**Story 3: Different Guarantees for Queue-Jumpers vs First-in-Line Jobs**

As a cluster administrator, I want workloads that were first in line to get a longer guaranteed runtime than workloads that jumped the queue by fitting into gaps.

Our ML team shares a queue with 8 GPUs. Large training jobs (8 GPUs) often wait while smaller experiment jobs (1-2 GPUs) get admitted into remaining capacity. With a single minAdmitDuration, we can't treat these cases differently. With two durations, we give first-in-line jobs 2 hours of guaranteed runtime while small jobs that jumped ahead only get 20 minutes — enough to finish or checkpoint, but not long enough to block the queue.

Comment (mukund-wayve) [reply keps/8522-time-based-same-priority-preemption/README.md:77]: This differentiation allows `minAdmitDuration` to be extended to lower priority preemption as well which is a requirement for us.

So, essentially, rather than differentiating between equal priority or lower priority, the differentiation comes from whether preemption is for workloads were admitted AFTER the HOL workload or BEFORE and it applies to lower or same priority. 

Maybe this KEP only targets the AFTER case and the BEFORE case is a new requirement I'm adding?

Comment (mukund-wayve) [reply keps/8522-time-based-same-priority-preemption/README.md:77]: I've created a [prototype PR](https://github.com/kubernetes-sigs/kueue/pull/9874/changes#diff-42d1c79488d61febf1bcdea04e7e78ff45c666a98325a309e17ec32f1d40b147R70).

Two differences from the KEP as written:

1. Two durations instead of one. Instead of a single minAdmitDuration, the prototype has minAdmitDurationSeconds for incumbent workloads (queued before the preemptor) and opportunisticMinAdmitDurationSeconds for workloads that jumped ahead (queued after the preemptor but admitted first, e.g. via BestEffortFIFO). This lets you give a longer guaranteed runtime to workloads that legitimately earned their spot vs. ones that fit opportunistically.

2. Works with **both policies**, not just LowerOrNewerEqualPriority. The KEP scopes withinClusterQueueConfig to LowerOrNewerEqualPriority only. The prototype applies the time check as a generic guard on top of both LowerPriority and LowerOrNewerEqualPriority -- since opportunistic admissions can happen under LowerPriority too with BestEffortFIFO. Each policy still decides which candidates are eligible using its existing rules; the time check just adds a minimum runtime before an eligible candidate can actually be preempted.

The alternative would be separate min-runtime settings per priority relationship (equal vs. lower), but that feels like too many options. The incumbent/opportunistic split covers the main use case more naturally.

I can raise a separate KEP for this if you want to avoid scope creep. But just want to highlight that this is a valid requirement for us and we might want to consider this generic setting.

Let me know what you think.

File: keps/8522-time-based-same-priority-preemption/README.md [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+## Proposal
+
+Introduce a new configuration structure `WithinClusterQueueConfig` that extends
+the `withinClusterQueue` preemption policy with a `minAdmitDuration` field.
+When `withinClusterQueue` is set to `LowerOrNewerEqualPriority` and
+`minAdmitDuration` is configured, workloads become eligible for same-priority
+preemption only after they have been admitted for at least the specified
+duration.
+
+### User Stories
```
Comment (olekzabl) [inline keps/8522-time-based-same-priority-preemption/README.md:100]: I can't help feeling that Story 1 is just a special case of Story 2.
(Where "resource" boils down to "GPU", "team" boils down to a "single researcher" etc.)

I'm curious, are there some differences or nuances which I've missed?

File: keps/8522-time-based-same-priority-preemption/README.md [hunk 4]
```diff
@@ -0,0 +1,353 @@
...
+- No single job can block others indefinitely
+- Resource access rotates fairly among all pending workloads
+
+### Notes/Constraints/Caveats (Optional)
+
+The `minAdmitDuration` field modifies the behavior of the
+`LowerOrNewerEqualPriority` policy. Without `minAdmitDuration`, the policy
+only preempts workloads that are newer than the preempting workload. With
+`minAdmitDuration`, it additionally preempts workloads that have exceeded
+the minimum duration, regardless of relative age.
```
Comment (olekzabl) [inline keps/8522-time-based-same-priority-preemption/README.md:143]: Nit, optional.
I understand it might feel like "irritating repetition", though I'm feeling for me it still helps.

```suggestion
The `minAdmitDuration` field modifies the behavior of the
`LowerOrNewerEqualPriority` policy. Without `minAdmitDuration`, the policy
only preempts (within same priority) workloads that are newer than the preempting workload. With
`minAdmitDuration`, it additionally preempts same-priority workloads that have exceeded
the minimum duration, regardless of relative age.
```

File: keps/8522-time-based-same-priority-preemption/README.md [hunk 5]
```diff
@@ -0,0 +1,353 @@
...
+This feature is complementary to Fair Sharing (KEP-1714). Fair Sharing
+addresses resource distribution across different ClusterQueues/Cohorts via
+preemption, while this feature addresses time-sharing among workloads with
+identical priority within the same ClusterQueue.
+
+**Interaction with Fair Sharing**: When both Fair Sharing and time-based
+preemption are enabled, time-based preemption only applies to the
+`withinClusterQueue` policy. Cross-ClusterQueue preemption continues to be
+governed by Fair Sharing algorithms. The two mechanisms operate independently
+and do not conflict.
```
Comment (olekzabl) [inline keps/8522-time-based-same-priority-preemption/README.md:154]: Does it feel like a good place to mention the proposed #9877 ?

File: keps/8522-time-based-same-priority-preemption/README.md [hunk 6]
```diff
@@ -0,0 +1,353 @@
...
+  - [Dynamic Priority Decay](#dynamic-priority-decay)
+  - [Rely on Admission Fair Sharing Alone](#rely-on-admission-fair-sharing-alone)
+  - [Usage-Based Preemption](#usage-based-preemption)
+<!-- /toc -->
+
+## Summary
+
+This KEP introduces time-based same-priority preemption, allowing workloads to
+preempt other workloads of equal priority after the target workloads have been
+admitted for a minimum duration (`minAdmitDuration`). This addresses the
```
Comment (olekzabl) [inline keps/8522-time-based-same-priority-preemption/README.md:38]: I'm sensing a clash of concepts:

- on one hand, you have `minAdmitDuration`, you also say "_admitted_ for a minimum duration", "_guaranteed runtime_" etc.
- OTOH, you propose "_quota reservation_ time" as the moment of "starting the watch".


IIUC this may significantly diverge e.g. if the CQ has multiple ResourceFlavors, _some_ of them based on a long-running AdmissionCheck (e.g. ProvisioningRequest; I heard these may take _hours_).

So in such case, we could set `minAdmitDuration: 4h` but still have a (same-priority) workload preempted e.g. after 3h from becoming _admitted_.

WDYT?

File: keps/8522-time-based-same-priority-preemption/README.md [hunk 7]
```diff
@@ -0,0 +1,353 @@
...
+    // withinClusterQueueConfig provides additional configuration for the
+    // withinClusterQueue preemption policy. Only valid when withinClusterQueue
+    // is set to LowerOrNewerEqualPriority.
+    // +optional
+    WithinClusterQueueConfig *WithinClusterQueueConfig `json:"withinClusterQueueConfig,omitempty"`
+}
+
+type WithinClusterQueueConfig struct {
+    // minAdmitDuration specifies the minimum duration a workload must be
+    // admitted before it becomes eligible for same-priority preemption.
```
Comment (olekzabl) [inline keps/8522-time-based-same-priority-preemption/README.md:208]: IIUC this sentence (especially with "_must_") is misleading because you're going to still allow preempting **newer** same-priority workloads, **even if** their "time since admission" (or even "time since quota reservation") is **below** `minAdmitDuration`. Am I right?

File: keps/8522-time-based-same-priority-preemption/README.md [hunk 8]
```diff
@@ -0,0 +1,353 @@
...
+
+#### Integration tests
+
+1. Two same-priority workloads competing for quota, first admitted workload
+   is preempted after `minAdmitDuration`
+2. Multiple workloads queued, fair rotation occurs as each exceeds threshold
+3. Mixed priorities: lower priority always preempted before same-priority
+4. `minAdmitDuration` not set: existing behavior preserved
+5. Edge case: workload exactly at `minAdmitDuration` threshold (should NOT
+   be preempted, as comparison requires strictly greater than)
```
Comment (olekzabl) [inline keps/8522-time-based-same-priority-preemption/README.md:284]: Is it worth it? I'm afraid this might be noticeable effort with minor value.

* Why minor value: given that the logic formula is
  `now - target.quotaReservationTime > minAdmitDuration`,
  where `now` is (I guess) some timestamp inside Kueue scheduler, it seems a super-rare event to hit `now` being _exactly equal_ to `quotaReservationTime + minAdmitDuration`.
  And, even if it happens, it seems super-irrelevant if we apply `>` or `>=` in that case :)
  Because, alternatively, we could easily have hit a timestamp 1ms earlier or later - so "either outcome feels legit" anyway.

* Why noticeable effort: to write a test for handling the `==` case, you'd need to inject some fake clock into the scheduler code.
  I currently see it done in unit tests but not in integration tests. (There, I only see fake clock injected into `kueuectl`).
  This _may_ be doable - but I'd blindly guess some effort will hide in that.

File: keps/8522-time-based-same-priority-preemption/README.md [hunk 9]
```diff
@@ -0,0 +1,353 @@
...
+6. Controller restart: workload timestamps persist correctly and preemption
+   timing remains accurate
+7. Multiple workloads expiring simultaneously: deterministic ordering based
+   on longest-running first
```
Comment (olekzabl) [inline keps/8522-time-based-same-priority-preemption/README.md:288]: I don't understand.
Is "moment of expiration" defined as "start time + `minAdmitDuration`", the latter summand independent of the workload?
If so, "expiring _simultaneously_" implies "they're all equally long-running".

My guess is that you don't need them to _expire simultaneously_; perhaps you just need a group of _expired_ workloads to be _simultaneously considered_ as preemption candidates for a single newly-incoming workload.

WDYT?

File: keps/8522-time-based-same-priority-preemption/README.md [hunk 10]
```diff
@@ -0,0 +1,353 @@
...
+
+- Feature behind `TimeBasedSamePriorityPreemption` feature gate (disabled by
+  default)
+- `withinClusterQueueConfig.minAdmitDuration` API field implemented
+- Preemption logic updated to consider time-expired workloads
+- Webhook validation for minimum duration and policy compatibility
+- Unit tests for preemption candidate selection
+- Integration tests for time-based preemption scenarios
+- Add a new reason value `InClusterQueueTimeBased` to the existing `reason`
+  label on `kueue_preempted_workloads_total` metric
```
Comment (olekzabl) [inline keps/8522-time-based-same-priority-preemption/README.md:306]: BTW how about some info in WorkloadStatus (I guess, specifically Conditions?)

Basically: how are we going to help a user understand "why was this workload preempted, even though the rules for `LowerOrNewerEqualPriority` which I used to know did not allow that"?

File: keps/8522-time-based-same-priority-preemption/README.md [hunk 11]
```diff
@@ -0,0 +1,353 @@
...
+
+### Preemption Eligibility
+
+Preemption only occurs when there is a pending workload that cannot be
+admitted due to insufficient quota. A pending workload can preempt an
+admitted same-priority workload if **either** of these conditions is true:
+
+1. **Existing behavior**: The target's quota reservation time is more recent
+   than the preemptor's creation time (i.e., the target was admitted after the
+   preemptor entered the queue)
```
Comment (olekzabl) [inline keps/8522-time-based-same-priority-preemption/README.md:227]: I don't think this description is accurate.
By comparing target's _quota reservation_ time with preemptor's _creation_ time, you postulate some asymmetry between target & preemptor which I don't see in the code.
(Neither [now](https://github.com/kubernetes-sigs/kueue/blob/baf8a94160fe65077c17a1387c57055c6309ac01/pkg/scheduler/preemption/common/preemption_policy.go#L39-L41) nor [before #9835](https://github.com/kubernetes-sigs/kueue/blob/22f2c7de247ff88fd14a7d88deaff57e068a0df6/pkg/scheduler/preemption/common/preemption_policy.go#L34-L36)).

File: keps/8522-time-based-same-priority-preemption/README.md [hunk 12]
```diff
@@ -0,0 +1,353 @@
...
+   `now - target.quotaReservationTime > minAdmitDuration`)
+
+Note: Condition 1 preserves backward compatibility with the existing
+`LowerOrNewerEqualPriority` behavior where a workload that has been waiting
+longer in the queue has priority over one that was admitted after the waiting
+workload was created.
+
+**Preemption order among same-priority candidates**:
+1. Time-expired workloads (exceeded `minAdmitDuration`) are preempted before
+   "newer" workloads (admitted after preemptor was created)
```
Comment (olekzabl) [inline keps/8522-time-based-same-priority-preemption/README.md:238]: I'd like to ask about this choice.
It doesn't feel _wrong_ to me - but it feels _not obvious_.

IIUC if we chose the other way (preempt "newer" before "time-expired" ones), we would also achieve the goals as outlined in the User Stories (i.e. avoid starvation of newer workloads), just in a more "conservative" way (i.e. remaining closer to the current semantics).

I'm sensing there should be noticeable difference between the two choices once we consider more complex scenarios, more workloads queueing over time etc.

I'd be happy to learn your thoughts on this.

**Review by olekzabl — COMMENTED** [review]
This seems to noticeably overlap with #9835 ...
At first glance, it feels that both efforts could be somehow merged?

cc @mbobrovskyi

File: keps/8522-time-based-same-priority-preemption/README.md (hunk 3 above)
Comment (sohankunkerkar) [reply keps/8522-time-based-same-priority-preemption/README.md:100]: They serve different audiences. Story 1 is a single-team GPU sharing scenario, Story 2 is a multi-team platform scenario. Keeping both for clarity.

File: keps/8522-time-based-same-priority-preemption/README.md (hunk 6 above)
Comment (sohankunkerkar) [reply keps/8522-time-based-same-priority-preemption/README.md:38]: @olekzabl Good catch! Looking at the code, the existing `LowerOrNewerEqualPriority`comparison uses `GetQueueOrderTimestamp()` for both preemptor and candidate, which returns `CreationTimestamp` by default, not `QuotaReserved` time. The KEP's use of "quota reservation time" in the terminology section was inaccurate about the existing behavior.
                                                                                                                                                                                                              For the new time-based path, the question is what timestamp to use for measuring "how long has this workload been running." I'll use the Admitted condition's `LastTransitionTime` because that's when pods actually started. This is a new comparison that doesn't interact with the existing ordering timestamp. Will fix the terminology section.

File: keps/8522-time-based-same-priority-preemption/README.md (hunk 7 above)
Comment (sohankunkerkar) [reply keps/8522-time-based-same-priority-preemption/README.md:208]: Will fix. `minAdmitDuration` adds a second eligibility path for older workloads. The existing newer path (creation-time-based, not reservation-time-based as the KEP incorrectly stated) is unchanged.

File: keps/8522-time-based-same-priority-preemption/README.md (hunk 8 above)
Comment (sohankunkerkar) [reply keps/8522-time-based-same-priority-preemption/README.md:284]: Agreed with your own assessment. Using >=, dropping the edge case test.

File: keps/8522-time-based-same-priority-preemption/README.md (hunk 11 above)
Comment (sohankunkerkar) [reply keps/8522-time-based-same-priority-preemption/README.md:227]: Yes, will fix that.

File: keps/8522-time-based-same-priority-preemption/README.md (hunk 12 above)
Comment (sohankunkerkar) [reply keps/8522-time-based-same-priority-preemption/README.md:238]: If we preempted newer first, time-expired workloads would never rotate out as there's almost always a newer candidate available. The long-running monopolizer keeps running, which is the exact problem this KEP solves. Preempting expired first ensures workloads that had their guaranteed runtime rotate out before those still within their protected period.

File: keps/8522-time-based-same-priority-preemption/README.md (hunk 9 above)
Comment (sohankunkerkar) [reply keps/8522-time-based-same-priority-preemption/README.md:288]: I will rewrite to `multiple expired candidates in the same scheduling cycle`.

File: keps/8522-time-based-same-priority-preemption/README.md (hunk 10 above)
Comment (sohankunkerkar) [reply keps/8522-time-based-same-priority-preemption/README.md:306]: Will add `InClusterQueueTimeBased` reason to the Evicted condition.

File: keps/8522-time-based-same-priority-preemption/README.md (hunk 2 above)
Comment (sohankunkerkar) [reply keps/8522-time-based-same-priority-preemption/README.md:77]: @mukund-wayve Agreed on including the two-duration model (minAdmitDuration + opportunisticMinAdmitDuration). The incumbent vs opportunistic distinction makes sense and the implementation is minimal. One thing to flag: @mimowo suggested on #8522 that the priority boost mechanism (PR #9120, merged) could achieve this via an external controller. PR #9959 is an example controller for it. However, priority  
  boost is per-workload and async. It can't do pairwise decisions (incumbent vs opportunistic), and lowering effective priority breaks out of same-priority semantics into the lowerPriority path in SatisfiesPreemptionPolicy, losing the fairness guarantee. Guaranteed minimum runtime needs a native check in the preemption eligibility path.                                                                                            
                                                                                                                                                                                                                                                             I still think the native minAdmitDuration approach is cleaner for admins (one config field vs deploying a controller).

Comment (mimowo) [reply keps/8522-time-based-same-priority-preemption/README.md:77]: > However, priority
boost is per-workload and async. It can't do pairwise decisions (incumbent vs opportunistic), and lowering effective priority breaks out of same-priority semantics into the lowerPriority path in SatisfiesPreemptionPolicy, losing the fairness guarantee. Guaranteed minimum runtime needs a native check in the preemption eligibility path.

@sohankunkerkar can you elaborate a bit more on that? I would like to understand what are the issues with the priority-boost round-robin before we go into full blown KEP for time-sharing. 

Yes, it is async, but it does not seem like a blocker on its own. The potential millisecond or even second delay for the "leader" swap does not seem prohibitive, because IIUC the time sharing we are talking here is 30min units anyway.

Comment (sohankunkerkar) [reply keps/8522-time-based-same-priority-preemption/README.md:77]: @mimowo You're right on the async part. It's not a concern for 30m+ windows. And I want to correct my earlier comment: the example controller uses a positive boost during the protection window, so it does work  for same-priority protection in the single-duration case.
                                                                                                                                                                                                                   Where the controller gets awkward is @mukund-wayve's use case: different protection for incumbents (2h) vs queue-jumpers (20m). To do that, the controller would need to list pending workloads, compare each  admitted workload's admission time against each pending workload's creation time, and classify incumbent vs opportunistic on every reconcile. It's reconstructing pairwise queue history that already exists inside `SatisfiesPreemptionPolicy` at preemption time. So for the single-duration case, the controller works, agreed. For the two-duration case with pairwise semantics, a native eligibility check is a more natural fit. @mukund-wayve can fill in more information here.

Comment (mukund-wayve) [reply keps/8522-time-based-same-priority-preemption/README.md:77]: Let me go back to one of our [use-cases](https://github.com/kubernetes-sigs/kueue/pull/8646/changes#r2934869206):

---

Story 3: Different Guarantees for Queue-Jumpers vs First-in-Line Jobs

As a cluster administrator, I want workloads that were first in line to get a longer guaranteed runtime than workloads that jumped the queue by fitting into gaps.

Our ML team shares a queue with 8 GPUs. Large training jobs (8 GPUs) often wait while smaller experiment jobs (1-2 GPUs) get admitted into available capacity via BestEffortFIFO. With a single minAdmitDuration, we can't treat these cases differently. With two durations, we give first-in-line jobs 2 hours of guaranteed runtime while small jobs that jumped ahead only get 20 minutes — enough to finish or checkpoint, but not long enough to block the queue.

---

Why I would prefer clear minRuntime semantics over priority boost (as a cluster admin):

1. It's much easier to reason about for users and cluster admins. Once a workloads reaches a certain threshold, it becomes eligible for preemption. This is easier to understand than relative priority boosts.
2. When one/more controllers are trying to change priorities for different reasons - time sharing, checkpointing, etc, it gets even more hard to reason how priorities are affecting preemptions.
3. No additional controller to manage


For these reasons, I would argue that min-runtime should be a first-class feature within Kueue.

**Comment by testinfected** [comment]
This is really valuable for us. @sohankunkerkar what are the next steps? Thanks for putting this up

**Comment by mimowo** [comment]
@sohankunkerkar do you want to continue on the KEP, or should we rather park it in favor of the PriorityBooster solution?

**Comment by sohankunkerkar** [comment]
The single-duration `minAdmitDuration` case folds naturally into #13396's MinExecutionTimeSeconds selector. The pairwise incumbent vs queue-jumper semantics that came out of this discussion are already tracked in #9876 (@mukund-wayve's KEP). Closing this PR. Thanks all for the discussions!
